README.md
# react-contacts
contacts  &amp; department select

[中文文档](https://github.com/wafersystems/react-contacts/blob/master/README_zh.md)

威发微服务选人组件/Wafer microservice selection component

[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/wafersystems/react-contacts)
[![npm](https://img.shields.io/npm/v/react-contacts.svg)](https://www.npmjs.com/package/react-contacts)
[![NPM downloads](https://img.shields.io/npm/dm/react-contacts.svg)](https://www.npmjs.com/package/react-contacts)

## Example

![Example](./example.png)

## How to use

### install

`yarn add react-contacts`

### React 

```js
import Contacts from 'react-contacts'

<Contacts {...props}/>

```

Properties  | Description                                                                                                         | Type | Default Values
------------- |---------------------------------------------------------------------------------------------------------------------| --------------| ------------- 
deptTree  | Department of Tree（[Department Data](###Department Data)）                                                           | array | []
users  | User data （[User & Search Result Data](###User & Search Result Data)）                                               | object | { records: []}
loading | Loading status                                                                                                      | bool | false
searchResult | Query user data return results （[User & Search Result Data](###User & Search Result Data)）                          | object | { records: []}
handleSearchUser | Handle search user function                                                                                         | func | function(page,nameKey,depId)
deptSearch | Show department search input                                                                                        | bool | tree
updateSelectUsers | Update user list when select user                                                                                   | func | function(userSelected)
updateSelectDept | Update dept list when select dept                                                                                   | func | function(deptSelected)
deptCheckBox | Show department checkbox                                                                                            | bool | true
searchDeptPlaceholder | search department placeholder                                                                                       | string | '请输入搜索部门'
searchUserPlaceholder | search department placeholder                                                                                       | string | '请输入搜索姓名'
defaultUserSelected   | default user selected                                                                                               | array | [{userId:'abc',username:'CCC'}] 
defaultDeptSelected   | default user selected ([Note](###Department tree initialization data))                                              | array | [{id:1,name:'ABC'}] 
numberColor   | total number color                                                                                                  | sring | #1B9AFF 
selectAllText   | select all text                                                                                                     | string | 全选 
totalShowText   | totalShowText                                                                                                       | string | 共选择了$个
userNameKey | fetch user name and selected key                                                                                    | string | 'username'
deptNameKey | Take the department tree and select the name of the key                                                             | string | 'name' 
radio   | show radio on user selected                                                                                         | bool | false
radioShowText   | when user selected on radio, show tip text.                                                                         | string | '已经选择' 
checkStrictly   | 	Check treeNode precisely; parent treeNode and children treeNodes are not associated                                | bool | false
showAllDeptTags   | Whether Tags show all nodes                                                                                         | bool | false
returnReducedNode | Department tree data reduction mode ([Department tree initialization data](###Department tree initialization data)) | bool | false
showLeft | show Left Panel                                                                                                     | bool | true
loadData | 	Load data asynchronously                                                                                           | function(node) ([Dynamically load the department tree](###Dynamically load the department tree)) | false
enNameKey| 	English name key                                                                                                   | string | false
disableUsers| 	The user ID cannot be selected                                                                                     | array | []
disableDept| 	The department ID cannot be selected                                                                               | array | []


### Department Data

````
[
      {
        id: 1,
        parentId: 0,
        children: [
          {
            id: 3,
            parentId: 1,
            children: [
              {
                id: 4,
                parentId: 3,
                children: [
                  {
                    id: 5,
                    parentId: 4,
                    children: [],
                    name: '院校农信',
                  },
                ],
                name: '高新农信',
              },
            ],
            name: '潍坊农信',
          },
        ],
        name: '山东农信',
      },
  ]
````

### User & Search Result Data

```

 {
      records: [
        {
          userId: 1,
          username: 'admin',
          password: '$2a$10$QOfWxxFyAMmEEmnuw9UI/..1s4B4eF/u9PzE2ZaGO.ij9YfmcUy.u',
          salt: null,
          wxOpenid: 'o_0FT0uyg_H1vVy2H0JpSwlVGhWQ',
          qqOpenid: null,
          createTime: '2018-04-20 07:15:18',
          updateTime: '2019-03-12 16:04:42',
          mail: 'a@a.com',
          delFlag: '0',
          lockFlag: '0',
          phone: '17034642888',
          avatar: 'lengleng-0d2a7b025da14d8d93f595b3fa082d82.jpg',
          deptId: 1,
          tenantId: 1,
          deptName: '真理部',
          roleList: [
            {
              roleId: 1,
              roleName: '管理员',
              roleCode: 'ROLE_ADMIN',
              roleDesc: '管理员',
              dsType: 2,
              dsScope: '2',
              createTime: '2017-10-29 15:45:51',
              updateTime: '2018-12-26 14:09:11',
              delFlag: '0',
            },
          ],
        },
	   ......
      ],
      total: 11,
      size: 10,
      current: 1,
      searchCount: true,
      pages: 1,
    },

```

###Department tree initialization data

returnReducedNode controls the data format of the department tree, true for simplified mode, false (default) for full data.

###Dynamically load the department tree

```javascript
          loadData={({deptId, children})=>{
            return new Promise(resolve => {
              if (children) {
                resolve();
                return;
              }
              setTimeout(() => {
              // 通过 deptId 查询下层数据，返回后合并到树的数据源上
                resolve();
              }, 1000);
            });
          }}
```

####Lean mode
If a node on the tree, its child nodes are all selected, the returned data only contains the parent node but not the child nodes

####Full mode
If a node on the tree, its child nodes are all selected, the returned data will contain the parent node and child nodes

####Initialize the department tree
If the initial data cannot contain the relationship between parent and child nodes, the department tree will appear out of sync when deleting with tag. Solution:

1. When saving data, please save the complete data, do not delete the content in children.

2. The use of reduced mode data may affect the background calculation.

Choose one of the above methods.

### Development

````
$ git clone https://github.com/wafersystems/react-contacts.git
$ yarn
$ yarn start

````


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [1.2.5] - 2022-05-20

### Added

- ✨ feat(v1): 人员搜索内容清空时，触发搜索回调, 传空字符串

### Changed

- none

### Removed

- none

&nbsp

## [1.2.4] - 2022-01-20

### Added

- none

### Changed

- ✨ refactor(v1): 增加控制不可选部门props, 增加 disableDept 控制不可选部门props，

### Removed

- none

&nbsp

## [1.2.3] - 2021-12-07

### Added

- none

### Changed

- 💄 style(v1): UI改版样式调整

### Removed

- none

&nbsp

## [1.1.9] - 2021-09-09

### Added

- none

### Changed

- add disableUsers props.通过userId控制不可选择人员

### Removed

- none

&nbsp

## [1.1.8] - 2021-09-06

### Added

- none

### Changed

- add enNameKey props.控制显示英文名

### Removed

- none

&nbsp

## [1.1.7] - 2021-08-24

### Added

- none

### Changed

- modify 调整部门搜索隐藏时的样式

### Removed

- none

&nbsp

## [1.1.6] - 2021-08-19

### Added

- none

### Changed

- modify 调整部门搜索隐藏时的样式

### Removed

- none

&nbsp

## [1.1.5] - 2021-08-19

### Added

- none

### Changed

- add props loadData, Load data asynchronously. 

### Removed

- none

&nbsp

## [1.1.4] - 2021-08-14

### Added

- none

### Changed

-  fix Default expansion error when the department tree has only one level.

### Removed

- none

&nbsp

## [1.1.3] - 2021-08-14

### Added

- none

### Changed

-  modify Naturally expand the first level of the department tree.

### Removed

- none

&nbsp

## [1.1.1] - 2021-08-05

### Added

- none

### Changed

-  fix 打开拖住排序显示Tag样式问题

### Removed

- none

&nbsp

## [1.1.0] - 2021-07-06

### Added

- none

### Changed

- modify 选中显示的结果样式提出

### Removed

- none

&nbsp

## [1.0.12] - 2021-06-30

### Added

- add Drag, 增加控制人员排序功能  

### Changed

- none

### Removed

- none

&nbsp

## [1.0.11] - 2021-06-21

### Added

- add props showLeft, 控制是否显示左边部门树面板，默认显示

### Changed

- none

### Removed

- none

&nbsp

## [1.0.10] - 2021-04-26

### Added

- none

### Changed

- modify 优化组件样式

### Removed

- none

&nbsp

## [1.0.9] - 2021-04-20

### Added

- none

### Changed

- Modify the scrollbar style under win.

### Removed

- none

&nbsp

## [1.0.8] - 2021-03-26

### Added

- none

### Changed

- modify list item hover color change to var(--list-item-hover-color)

### Removed

- none

&nbsp

## [1.0.5] - 2021-02-05

### Added

- none

### Changed

- modify result div border color change to var(--border-color)

### Removed

- none

&nbsp

## [1.0.4] - 2021-02-05

### Added

- none

### Changed

- modify result div background color change to var(--bg-color)

### Removed

- none

&nbsp

## [1.0.3] - 2020-11-26

### Added

- none

### Changed

- fix 默认查询时部门和nameKey缺失问题

### Removed

- none

&nbsp;

## [1.0.2] - 2021-01-08

### Added

- none

### Changed

- fix 部门过长显示问题

### Removed

- none

&nbsp;

## [1.0.1] - 2020-11-26

### Added

- none

### Changed

- fix Tree selected id must be string.

### Removed

- none

&nbsp;

## [1.0.0] - 2020-11-26

### Added

- none

### Changed

- update Antd to 4.x

### Removed

- none

&nbsp;


## [0.0.26] - 2020-11-06

### Added

- none

### Changed

- fix If showAllDeptTags is false, the count is wrong.

### Removed

- none

&nbsp;

## [0.0.26] - 2020-09-22

### Added

- none

### Changed

- add returnReducedNode props.Return the simplified node. If true, only return the simplified node. For example, if all child nodes are selected, only the parent node will be returned.

### Removed

- none

&nbsp;

## [0.0.25] - 2020-07-16

### Added

- none

### Changed

- fix Page does not update the department id problem.

### Removed

- none

&nbsp;

## [0.0.24] - 2020-05-28

### Added

- none

### Changed

- modify Controls whether tags display all nodes

### Removed

- none

&nbsp;

## [0.0.23] - 2020-05-28

### Added

- none

### Changed

- fix the problem of deleting the tree display from the tag.

### Removed

- none

&nbsp;

## [0.0.22] - 2020-05-28

### Added

- none

### Changed

- fix Department tree deselection problem fix.

### Removed

- none

&nbsp;

## [0.0.21] - 2020-05-27

### Added

- none

### Changed

- fix department selection problem in the selection component.

### Removed

- none

&nbsp;

## [0.0.20] - 2019-12-20

### Added

- none

### Changed

- modify Adjust the style.

### Removed

- none

&nbsp; 

## [0.0.19] - 2019-12-20

### Added

- none

### Changed

- modify pages current start with 1.

### Removed

- none

&nbsp; 

## [0.0.18] - 2019-12-02

### Added

- none

### Changed

- fix department search error & user search .

### Removed

- none

&nbsp; 

## [0.0.17] - 2019-12-02

### Added

- none

### Changed

- fix department search error.

### Removed

- none

&nbsp; 

## [0.0.16] - 2019-12-02

### Added

- none

### Changed

- fix when default selected all user,the all checkbox unchecked.
- fix when department changed，the all checkbox don't refresh.

### Removed

- none

&nbsp; 

## [0.0.15] - 2019-11-28

### Added

- none

### Changed

- modify when select all user, the all checkbox will be checked.

### Removed

- none

&nbsp; 

## [0.0.14] - 2019-11-28

### Added

- none

### Changed

- fix deptCheckBox can't control department search result show.
- fix select all clean all data issues.
- modify adjust test data.

### Removed

- none

&nbsp; 

## [0.0.13] - 2019-10-29

### Added

- none

### Changed

- rebuild dist.

### Removed

- none

&nbsp; 

## [0.0.12] - 2019-10-28

### Added

- add radio & radioShowText ,can set user selected to radio.

### Changed

- none

### Removed

- none

&nbsp; 

## [0.0.11] - 2019-10-25

### Added

- none

### Changed

- fix default user & dept not set to handle.

### Removed

- none

&nbsp; 

## [0.0.10] - 2019-10-24

### Added

- add props defaultDeptSelected set default selected dept.
- add props userNameKey to show user name.
- add props deptNameKey to show dept name.

### Changed

- none

### Removed

- none

&nbsp; 

## [0.0.9] - 2019-10-23


src/mockData.js
/**
 * Created : vincent
 * Date : 2019-07-10  16:57
 * Email : wangxiao@wafersystems.com
 */

export const deptTree = [{
  id: 1,
  parentId: 0,
  children: [
    {
      id: 3,
      parentId: 1,
      children: [
        {
          id: 4,
          parentId: 3,
          children: [
            {
              id: 5,
              parentId: 4,
              children: [],
              name: '院校农信',
            },
            {
              id: 6,
              parentId: 4,
              children: [],
              name: '院校农信123',
            },
          ],
          name: '高新农信',
        },
        {
          id: 14,
          parentId: 3,
          children: [
            {
              id: 15,
              parentId: 14,
              children: [],
              name: '院校农信B',
            },
            {
              id: 16,
              parentId: 14,
              children: [],
              name: '院校农信A',
            },
          ],
          name: '高新农信C',
        },
      ],
      name: '潍坊农信',
    },
  ],
  name: '山东农信',
}];

export const userList = {
  records: [
    {
      userId: 1,
      username: 'admin',
      password: '$2a$10$QOfWxxFyAMmEEmnuw9UI/..1s4B4eF/u9PzE2ZaGO.ij9YfmcUy.u',
      salt: null,
      wxOpenid: 'o_0FT0uyg_H1vVy2H0JpSwlVGhWQ',
      qqOpenid: null,
      createTime: '2018-04-20 07:15:18',
      updateTime: '2019-03-12 16:04:42',
      mail: 'a@a.com',
      delFlag: '0',
      lockFlag: '0',
      phone: '17034642888',
      avatar: 'lengleng-0d2a7b025da14d8d93f595b3fa082d82.jpg',
      deptId: 1,
      tenantId: 1,
      deptName: '真理部真理部真理部真理部真理部真理部真理部真理部真理部真理部真理部',
      roleList: [
        {
          roleId: 1,
          roleName: '管理员',
          roleCode: 'ROLE_ADMIN',
          roleDesc: '管理员',
          dsType: 2,
          dsScope: '2',
          createTime: '2017-10-29 15:45:51',
          updateTime: '2018-12-26 14:09:11',
          delFlag: '0',
        },
      ],
    },
    {
      userId: 11,
      username: 'admin1',
      password: '$2a$10$QOfWxxFyAMmEEmnuw9UI/..1s4B4eF/u9PzE2ZaGO.ij9YfmcUy.u',
      salt: null,
      wxOpenid: 'o_0FT0uyg_H1vVy2H0JpSwlVGhWQ',
      qqOpenid: null,
      createTime: '2018-04-20 07:15:18',
      updateTime: '2019-03-12 16:04:42',
      mail: 'a@a.com',
      delFlag: '0',
      lockFlag: '0',
      phone: '17034642888',
      avatar: 'lengleng-0d2a7b025da14d8d93f595b3fa082d82.jpg',
      deptId: 1,
      tenantId: 1,
      deptName: '真理部',
      roleList: [
        {
          roleId: 1,
          roleName: '管理员',
          roleCode: 'ROLE_ADMIN',
          roleDesc: '管理员',
          dsType: 2,
          dsScope: '2',
          createTime: '2017-10-29 15:45:51',
          updateTime: '2018-12-26 14:09:11',
          delFlag: '0',
        },
      ],
    },
    {
      userId: 111,
      username: 'admin11',
      password: '$2a$10$QOfWxxFyAMmEEmnuw9UI/..1s4B4eF/u9PzE2ZaGO.ij9YfmcUy.u',
      salt: null,
      wxOpenid: 'o_0FT0uyg_H1vVy2H0JpSwlVGhWQ',
      qqOpenid: null,
      createTime: '2018-04-20 07:15:18',
      updateTime: '2019-03-12 16:04:42',
      mail: 'a@a.com',
      delFlag: '0',
      lockFlag: '0',
      phone: '17034642888',
      avatar: 'lengleng-0d2a7b025da14d8d93f595b3fa082d82.jpg',
      deptId: 1,
      tenantId: 1,
      deptName: '真理部',
      roleList: [
        {
          roleId: 1,
          roleName: '管理员',
          roleCode: 'ROLE_ADMIN',
          roleDesc: '管理员',
          dsType: 2,
          dsScope: '2',
          createTime: '2017-10-29 15:45:51',
          updateTime: '2018-12-26 14:09:11',
          delFlag: '0',
        },
      ],
    },
    {
      userId: 1111,
      username: 'admin111',
      password: '$2a$10$QOfWxxFyAMmEEmnuw9UI/..1s4B4eF/u9PzE2ZaGO.ij9YfmcUy.u',
      salt: null,
      wxOpenid: 'o_0FT0uyg_H1vVy2H0JpSwlVGhWQ',
      qqOpenid: null,
      createTime: '2018-04-20 07:15:18',
      updateTime: '2019-03-12 16:04:42',
      mail: 'a@a.com',
      delFlag: '0',
      lockFlag: '0',
      phone: '17034642888',
      avatar: 'lengleng-0d2a7b025da14d8d93f595b3fa082d82.jpg',
      deptId: 1,
      tenantId: 1,
      deptName: '真理部',
      roleList: [
        {
          roleId: 1,
          roleName: '管理员',
          roleCode: 'ROLE_ADMIN',
          roleDesc: '管理员',
          dsType: 2,
          dsScope: '2',
          createTime: '2017-10-29 15:45:51',
          updateTime: '2018-12-26 14:09:11',
          delFlag: '0',
        },
      ],
    },
    {
      userId: 11111,
      username: 'admin1111',
      password: '$2a$10$QOfWxxFyAMmEEmnuw9UI/..1s4B4eF/u9PzE2ZaGO.ij9YfmcUy.u',
      salt: null,
      wxOpenid: 'o_0FT0uyg_H1vVy2H0JpSwlVGhWQ',
      qqOpenid: null,
      createTime: '2018-04-20 07:15:18',
      updateTime: '2019-03-12 16:04:42',
      mail: 'a@a.com',
      delFlag: '0',
      lockFlag: '0',
      phone: '17034642888',
      avatar: 'lengleng-0d2a7b025da14d8d93f595b3fa082d82.jpg',
      deptId: 1,
      tenantId: 1,
      deptName: '真理部',
      roleList: [
        {
          roleId: 1,
          roleName: '管理员',
          roleCode: 'ROLE_ADMIN',
          roleDesc: '管理员',
          dsType: 2,
          dsScope: '2',
          createTime: '2017-10-29 15:45:51',
          updateTime: '2018-12-26 14:09:11',
          delFlag: '0',
        },
      ],
    },
    {
      userId: 111111,
      username: 'admin11111',
      password: '$2a$10$QOfWxxFyAMmEEmnuw9UI/..1s4B4eF/u9PzE2ZaGO.ij9YfmcUy.u',
      salt: null,
      wxOpenid: 'o_0FT0uyg_H1vVy2H0JpSwlVGhWQ',
      qqOpenid: null,
      createTime: '2018-04-20 07:15:18',
      updateTime: '2019-03-12 16:04:42',
      mail: 'a@a.com',
      delFlag: '0',
      lockFlag: '0',
      phone: '17034642888',
      avatar: 'lengleng-0d2a7b025da14d8d93f595b3fa082d82.jpg',
      deptId: 1,
      tenantId: 1,
      deptName: '真理部',
      roleList: [
        {
          roleId: 1,
          roleName: '管理员',
          roleCode: 'ROLE_ADMIN',
          roleDesc: '管理员',
          dsType: 2,
          dsScope: '2',
          createTime: '2017-10-29 15:45:51',
          updateTime: '2018-12-26 14:09:11',
          delFlag: '0',
        },
      ],
    },
    {
      userId: 1111111,
      username: 'admin111111',
      password: '$2a$10$QOfWxxFyAMmEEmnuw9UI/..1s4B4eF/u9PzE2ZaGO.ij9YfmcUy.u',
      salt: null,
      wxOpenid: 'o_0FT0uyg_H1vVy2H0JpSwlVGhWQ',
      qqOpenid: null,
      createTime: '2018-04-20 07:15:18',
      updateTime: '2019-03-12 16:04:42',
      mail: 'a@a.com',
      delFlag: '0',
      lockFlag: '0',
      phone: '17034642888',
      avatar: 'lengleng-0d2a7b025da14d8d93f595b3fa082d82.jpg',
      deptId: 1,
      tenantId: 1,
      deptName: '真理部',
      roleList: [
        {
          roleId: 1,
          roleName: '管理员',
          roleCode: 'ROLE_ADMIN',
          roleDesc: '管理员',
          dsType: 2,
          dsScope: '2',
          createTime: '2017-10-29 15:45:51',
          updateTime: '2018-12-26 14:09:11',
          delFlag: '0',
        },
      ],
    },
    {
      userId: 11111111,
      username: 'admin1111111',
      password: '$2a$10$QOfWxxFyAMmEEmnuw9UI/..1s4B4eF/u9PzE2ZaGO.ij9YfmcUy.u',
      salt: null,
      wxOpenid: 'o_0FT0uyg_H1vVy2H0JpSwlVGhWQ',
      qqOpenid: null,
      createTime: '2018-04-20 07:15:18',
      updateTime: '2019-03-12 16:04:42',
      mail: 'a@a.com',
      delFlag: '0',
      lockFlag: '0',
      phone: '17034642888',
      avatar: 'lengleng-0d2a7b025da14d8d93f595b3fa082d82.jpg',
      deptId: 1,
      tenantId: 1,
      deptName: '真理部',
      roleList: [
        {
          roleId: 1,
          roleName: '管理员',
          roleCode: 'ROLE_ADMIN',
          roleDesc: '管理员',
          dsType: 2,
          dsScope: '2',
          createTime: '2017-10-29 15:45:51',
          updateTime: '2018-12-26 14:09:11',
          delFlag: '0',
        },
      ],
    },
    {
      userId: 111111111,
      username: 'admin11111111',
      password: '$2a$10$QOfWxxFyAMmEEmnuw9UI/..1s4B4eF/u9PzE2ZaGO.ij9YfmcUy.u',
      salt: null,
      wxOpenid: 'o_0FT0uyg_H1vVy2H0JpSwlVGhWQ',
      qqOpenid: null,
      createTime: '2018-04-20 07:15:18',
      updateTime: '2019-03-12 16:04:42',
      mail: 'a@a.com',
      delFlag: '0',
      lockFlag: '0',
      phone: '17034642888',
      avatar: 'lengleng-0d2a7b025da14d8d93f595b3fa082d82.jpg',
      deptId: 1,
      tenantId: 1,
      deptName: '真理部',
      roleList: [
        {
          roleId: 1,
          roleName: '管理员',
          roleCode: 'ROLE_ADMIN',
          roleDesc: '管理员',
          dsType: 2,
          dsScope: '2',
          createTime: '2017-10-29 15:45:51',
          updateTime: '2018-12-26 14:09:11',
          delFlag: '0',
        },
      ],
    },
    {
      userId: 101,
      username: 'admin01',
      password: '$2a$10$QOfWxxFyAMmEEmnuw9UI/..1s4B4eF/u9PzE2ZaGO.ij9YfmcUy.u',
      salt: null,
      wxOpenid: 'o_0FT0uyg_H1vVy2H0JpSwlVGhWQ',
      qqOpenid: null,
      createTime: '2018-04-20 07:15:18',
      updateTime: '2019-03-12 16:04:42',
      mail: 'a@a.com',
      delFlag: '0',
      lockFlag: '0',
      phone: '17034642888',
      avatar: 'lengleng-0d2a7b025da14d8d93f595b3fa082d82.jpg',
      deptId: 1,
      tenantId: 1,
      deptName: '真理部',
      roleList: [
        {
          roleId: 1,
          roleName: '管理员',
          roleCode: 'ROLE_ADMIN',
          roleDesc: '管理员',
          dsType: 2,
          dsScope: '2',
          createTime: '2017-10-29 15:45:51',
          updateTime: '2018-12-26 14:09:11',
          delFlag: '0',
        },
      ],
    },
  ],
  total: 11,
  size: 10,
  current: 1,
  searchCount: true,
  pages: 1,
};

export const u2 = {
  records: [],
  total: 0,
  size: 10,
  current: 1,
  searchCount: true,
  pages: 1,
}
